src/binary-encoding.adoc

```diff
@@ -0,0 +1,1618 @@
+
+[[delimited_structs]]
+==== Delimited structs
+
+Opcode `0xF3` indicates the beginning of a delimited struct with <<varsym, `VarSym`>> field names,
+while opcode `0xF0` closes the most recently opened delimited container that has not yet been closed.
+
+NOTE: While length-prefixed structs can choose between <<structs_with_symbol_address_field_names, structs with
+symbol address field names>> and <<structs_with_varsym_field_names, structs with `VarSym` field names>>,
+delimited structs always use `VarSym`-encoded field names.
+
+===== Example encoding of the empty struct (`{}`)
+[source]
+----
+┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.│
+│  ┌─── VarSym escape code 0x00: an opcode follows
+│  │  ┌─── Opcode 0xF0 indicates the end of the most
+│  │  │    recently opened delimited container
+F3 00 F0
+----
+
+===== Example encoding of `{"foo": 1, $11: 2}`
+[source]
+----
+┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.
+│
+│  ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
+│  │                 │        ┌─── VarSym escape code 0x00: an opcode follows
+│  │                 │        │  ┌─── Opcode 0xF0 indicates the end of the most
+│  │   f  o  o       │        │  │    recently opened delimited container
+F3 FD 66 6F 6F 51 01 17 91 02 00 F0
+      └──┬───┘ └─┬─┘    └─┬─┘
+      3 UTF-8    1        2
+       bytes
+----
+
+[[nulls]]
+=== Nulls
+
+The opcode `0xEA` indicates an untyped null (that is: `null`, or its alias `null.null`).
+
+The opcode `0xEB` indicates a typed null; a byte follows whose value indicates the type. The byte-to-type
+mapping is as follows:
+
+[cols="^1a,4a"]
+|===
+|Byte |Type
+
+|`0x00`
+|`null.bool`
+
+|`0x01`
+|`null.int`
+
+|`0x02`
+|`null.float`
+
+|`0x03`
+|`null.decimal`
+
+|`0x04`
+|`null.timestamp`
+
+|`0x05`
+|`null.string`
+
+|`0x06`
+|`null.symbol`
+
+|`0x07`
+|`null.clob`
+
+|`0x08`
+|`null.blob`
+
+|`0x09`
+|`null.list`
+
+|`0x0A`
+|`null.sexp`
+
+|`0x0B`
+|`null.struct`
+|===
+
+==== Example encoding of `null`
+[source]
+----
+┌──── The opcode `0xEA` represents a null (null.null)
+EA
+----
+
+==== Example encoding of `null.string`
+[source]
+----
+┌──── The opcode `0xEB` indicates a typed null; a byte indicating the type follows
+│  ┌──── Byte 0x05 indicates the type `string`
+EB 0x5
+----
+
+[[annotations]]
+=== Annotations
+
+Annotations can be encoded either <<annotations_with_symbol_addresses, as symbol addresses>>
+or <<annotations_with_varsym_text, as ``VarSym``s>>. In both encodings, the annotations sequence appears
+just before the value that it decorates.
+
+If an annotations sequence appears before one or more additional annotations sequences, the sequences
+are concatenated.
+
+It is illegal for an annotations sequence to appear before any of the following:
+
+* The end of the stream
+* A <<nops, `NOP`>>
+// TODO: Links
+* An E-expression (that is: a macro invocation). To add annotations to a macro invocation, see the
```

Comment:
I just realized that in the table *and* the macro invocation sections we don't call out these things as E-Expressions, I think we should be clearer about that there.

Reply:
Good call, done.